academy/3-ibc/solomachine.md

```diff
@@ -0,0 +1,378 @@
+
+``` typescript
+interface ClientState {
+  frozen: boolean
+  consensusState: ConsensusState
+}
+```
+
+It must of course also implement the `ClientState` methods defined in [the spec](https://github.com/cosmos/ibc/tree/main/spec/core/ics-002-client-semantics#clientstate).
+
+#### `ConsensusState`
+
+Next to the public key, the `ConsensusSate` also contains a timestamp and diversifier. The diversifier is an arbitrary string, chosen when the client is created, designed to allow the same public key to be re-used across different solo machine clients (potentially on different chains) without being considered misbehavior.
+
+```typescript
+interface ConsensusState {
+  sequence: uint64 // deprecated
+  publicKey: PublicKey
+  diversifier: string
+  timestamp: uint64
+}
+```
+
+It must of course also implement the `ConsensusState` methods defined in [the spec](https://github.com/cosmos/ibc/tree/main/spec/core/ics-002-client-semantics#consensusstate).
+
+### Verifying signatures
+
+The essence of the solomachine / solomachine client pair is that we can verify signatures based on public/private key cryptography. The public key is stored in the `ConsensusState` of the solomachine client and can be used to verify signatures signed with the solomachine private key corresponding to the available public key. This signature verification is used in the following situations:
+
+* When updating the `ConsensusState` with a new public key or diversifier through submitting a `Header`.
+* When submitting evidence of `Misbehaviour` to freeze a malicious client.
+* When verifying packet commitments.
+
+#### Signing data
+
+The solomachine will sign over some data, `SignBytes`:
+
+``` typescript
+interface SignBytes {
+  sequence: uint64
+  timestamp: uint64  
+  diversifier: string
+  path: []byte
+  data: []byte
+}
+```
+
+and store a `Signature` including the data bytes and a timestamp:
+
+``` typescript
+interface Signature {
+  data: []byte
+  timestamp: uint64
+}
+```
+
+This signature is then added to a message, e.g. the `Header`:
+
+```typescript
+interface Header {
+  sequence: uint64 // deprecated
+  timestamp: uint64
+  signature: Signature
+  newPublicKey: PublicKey
+  newDiversifier: string
+}
+```
+
+#### Client message signature verification
+
+Remember from the section on light client development that the client can be updated (with new state or evidence of misbehavior) by submitting a `ClientMessage`. 
+
+The `ClientMessage` will be passed onto the client through a `MsgUpdateClient` submitted (generally by a relayer). The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):
```

Comment:
```suggestion
The `ClientMessage` will be passed onto the client by submitting a `MsgUpdateClient`. Although IBC messages are typically submitted by relayers, it can be advantageous to integrate a relayer directly into the solomachine's server for certain cases, such as with crypto.com's implementation. The `02-client`'s [`UpdateClient`](https://github.com/cosmos/ibc-go/blob/v7.0.0/modules/core/02-client/keeper/client.go#L48) method will then handle the client message by using [these 4 methods on the `ClientState` interface](https://github.com/cosmos/ibc-go/blob/02-client-refactor-beta1/modules/core/exported/client.go#L98-L109):
```